collector/bcache.go

```diff
@@ -0,0 +1,320 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nobcache
+
+package collector
+
+import (
+	"fmt"
+
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/bcache"
+	"github.com/prometheus/procfs/sysfs"
+)
+
+func init() {
+	Factories["bcache"] = NewBcacheCollector
+}
+
+// A bcacheCollector is a Collector which gathers metrics from Linux bcache
+type bcacheCollector struct {
+	fs sysfs.FS
+}
+
+// NewBcacheCollector returns a newly allocated bcacheCollector.
+// It exposes a number of Linux bcache statistics.
+func NewBcacheCollector() (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %v", err)
+	}
+
+	return &bcacheCollector{
+		fs: fs,
+	}, nil
+}
+
+// Update reads and exposes bcache stats.
+// It implements the Collector interface.
+func (c *bcacheCollector) Update(ch chan<- prometheus.Metric) error {
+	stats, err := c.fs.BcacheStats()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve bcache stats: %v", err)
+	}
+
+	for _, s := range stats {
+		c.updateBcacheStats(ch, s)
+	}
+	return nil
+}
+
+type metric struct {
```

Comment:
We only have one global namespace in the collector package, please use unique types and function names like `bcacheMetric`.